libp2p/network/network.go

```diff
@@ -0,0 +1,219 @@
+	if selfPoktAddr.Equals(poktAddr) {
+		return nil
+	}
+
+	poktPeer, ok := p2pNet.addrBookMap[poktAddr.String()]
+	if !ok {
+		// NB: this should not happen.
+		return ErrNetwork("", fmt.Errorf(
+			"peer not found in address book, pokt address: %s", poktAddr,
+		))
+	}
+
+	peerAddrInfo, err := identity.PeerAddrInfoFromPoktPeer(poktPeer)
+	if err != nil {
+		return ErrNetwork("parsing peer multiaddr", err)
+	}
+
+	stream, err := p2pNet.host.NewStream(ctx, peerAddrInfo.ID, protocol.PoktProtocolID)
+	if err != nil {
+		return ErrNetwork(fmt.Sprintf(
+			"opening peer stream, pokt address: %s", poktAddr,
+		), err)
+	}
+
+	if _, err := stream.Write(data); err != nil {
+		return ErrNetwork(fmt.Sprintf(
+			"writing to stream (peer address: %s)", poktAddr,
+		), err)
+	}
+	defer func() {
+		if err := stream.Close(); err != nil {
+			p2pNet.logger.Error().Err(err)
+		}
+	}()
+
+	// TECHDEBT: check if there's a more conventional way to send
+	// messages directly than to close and re-open per direct message.
+	// NB: close the stream so that peer receives EOF.
+	if err := stream.Close(); err != nil {
+		return ErrNetwork(fmt.Sprintf(
+			"closing peer stream, pokt address: %s", poktAddr,
+		), err)
+	}
+	return nil
+}
+
+// This function was added to specifically support the RainTree implementation.
+// Handles the raw data received from the network and returns the data to be processed
+// by the application layer.
+func (p2pNet *libp2pNetwork) HandleNetworkData(data []byte) ([]byte, error) {
+	return data, nil
+}
+
+func (p2pNet *libp2pNetwork) GetAddrBook() types.AddrBook {
+	addrBook := make(types.AddrBook, 0)
+	for _, poktPeer := range p2pNet.addrBookMap {
+		addrBook = append(addrBook, poktPeer)
+	}
+	return addrBook
+}
+
+func (p2pNet *libp2pNetwork) AddPeerToAddrBook(poktPeer *types.NetworkPeer) error {
+	p2pNet.addrBookMap[poktPeer.Address.String()] = poktPeer
+
+	pubKey, err := identity.PubKeyFromPoktPeer(poktPeer)
+	if err != nil {
+		return ErrNetwork(fmt.Sprintf(
+			"converting peer public key, pokt address: %s", poktPeer.Address,
+		), err)
+	}
+	peer, err := identity.PeerAddrInfoFromPoktPeer(poktPeer)
+	if err != nil {
+		return ErrNetwork(fmt.Sprintf(
+			"converting peer info, pokt address: %s", poktPeer.Address,
+		), err)
+	}
+
+	p2pNet.host.Peerstore().AddAddrs(peer.ID, peer.Addrs, DefaultPeerTTL)
+	if err := p2pNet.host.Peerstore().AddPubKey(peer.ID, pubKey); err != nil {
+		return ErrNetwork(fmt.Sprintf(
+			"adding peer public key, pokt address: %s", poktPeer.Address,
+		), err)
+	}
+	return nil
+}
+
+// CLEANUP: fix typo in interface (?)
+func (p2pNet *libp2pNetwork) RemovePeerToAddrBook(poktPeer *types.NetworkPeer) error {
+	delete(p2pNet.addrBookMap, poktPeer.Address.String())
+
+	peer, err := identity.PeerAddrInfoFromPoktPeer(poktPeer)
+	if err != nil {
+		return ErrNetwork(fmt.Sprintf(
+			"converting peer info, pokt address: %s", poktPeer.Address,
+		), err)
+	}
+
+	p2pNet.host.Peerstore().RemovePeer(peer.ID)
```

Comment:
Is the peerstore a conn pool?

Reply:
Noted. Let's definitely build an understanding around these abstractions if we're going to be heavily dependant on them.